### PR TITLE
Composites Blog Post

### PR DESCRIPTION
**Pull request by christophermaier**

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: www/source/blog/2017-10-05-Introducing-Composites.html.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+The final new field, `pkg_bind_map`, is also the most exciting. We have already seen how we are connecting the `http` binding of our Nginx proxy to the exported `port` from our NodeJS application. However, there's no reason that you, the user, should always have to specify this binding on the command line when you start this composite up. It's basically an implementation detail of this particular arrangement of services. It's maybe not be that big of a deal in this particular example, but as you add more services with multiple binds, it really starts to become a challenge. That's where `pkg_bind_map` comes in. This lets you declare all the binding relationships between all the services in the composite. Here we are saying that our `christophermaier/composite-example-api-proxy` has a bind named `http` that is satisfied by the `christophermaier/sample-node-app` service. If our proxy had more than one bind, then we could add more mappings, separated by spaces, like so: `"http:christophermaier/sample-node-app otherbind:other/service-in-the-composite"`.
```
Comment (mwrock) [inline www/source/blog/2017-10-05-Introducing-Composites.html.md]: `It's maybe not be that big of a deal` -> `It's maybe not that big of a deal`

Comment (mwrock) [reply www/source/blog/2017-10-05-Introducing-Composites.html.md]: Also I love this functionality bringing in the bind args into a plan!

File: www/source/blog/2017-10-05-Introducing-Composites.html.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+A particularly fun part of composites is the fact that build process is intelligent. If you list a package in your `pkg_services` array that isn't actually a service (i.e., it doesn't have a run hook), it will trigger a build error. Additionally, you must also provide at least two services; after all, a composite with a single service isn't providing you anything that the service by itself can't. Finally, you cannot build a composite if it specifies bindings that are not possible. For example, our proxy's `http` bind requires an exported value of `port` to be satisfied (see the proxy's `plan.sh` file above). If our NodeJS application exported a value of `http_port` instead, the composite package's build would fail when it detects this mismatch. This is another example of Habitat's overall philosophy of surfacing errors to the user as quickly as possible. It's much better to discover these kinds of issues at build time than at deploy time.
```
Comment (mwrock) [inline www/source/blog/2017-10-05-Introducing-Composites.html.md]: `is the fact that build process` -> `is the fact that the build process`

File: www/source/blog/2017-10-05-Introducing-Composites.html.md [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+## What Composites _Aren't_
+
+Composite packages are a new feature of Habitat, and there may be a few rough edges. The current implementation is intended as a way to explore this problem space for Habitat, but it is not necessarily the _final_ implementation. We are very curious to see how composite packages work (or don't!) for you and your particular use cases.
+
+While you can install, start, stop, load, and unload composite packages, you should keep in mind that composite packages can behave differently from the Habitat packages you are used to. In particular, composite packages are not _quite_ a formal runtime construct of Habitat. The individual services are still available and addressable just as though you'd started them manually. The composite-as-a-whole is not directly addressable as though it were a service in its own right. It cannot not do anything like export selected values from its constituent services as "composite-level exports", or map "composite-level binds" onto constituent service binds. A composite package as a whole does not update itself as an individual service can. Indeed, the constituent services of a composite can continue to update individually.
```
Comment (mwrock) [inline www/source/blog/2017-10-05-Introducing-Composites.html.md]: `It cannot not do ` -> `It cannot do `

File: www/source/blog/2017-10-05-Introducing-Composites.html.md [hunk 4]
```diff
@@ -0,0 +1,274 @@
+---
+title: Introducing Composites
+date: 2017-10-05
```
Comment (reset) [inline www/source/blog/2017-10-05-Introducing-Composites.html.md]: Let's kick this date to today

File: www/source/blog/2017-10-05-Introducing-Composites.html.md [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+author: Christopher Maier
+tags: blog
```
Comment (reset) [inline www/source/blog/2017-10-05-Introducing-Composites.html.md]: You might want to copy the categories into this field, too, or it'll just be tagged as "blog" to the left pane of the blog

File: www/source/blog/2017-10-05-Introducing-Composites.html.md [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+## Why?
+
+In the past, the intrepid Habitat user community has been finding all sorts of creative ways to use Habitat in a variety of ways. Since the sidecar pattern is becoming increasingly prevalent, particularly with microservice architectures and containerized deployments, users have tried to implement it in Habitat, but with limited success. Often, we would see users trying to start multiple services from the `init` or `run` hook of another coordinating service. Approaches like this may work initially, but won't work in the long run, because the Habitat Supervisor won't be properly supervising additional services that are spawned this way.
```
Comment (reset) [inline www/source/blog/2017-10-05-Introducing-Composites.html.md]: I might link off to the sidecar pattern or an example of it here

Comment (reset) [reply www/source/blog/2017-10-05-Introducing-Composites.html.md]: Might be worth calling out here that any calls to `hab` is undefined behaviour in a hook

**Comment by christophermaier** [comment]
@thesentinels acceptance

**Comment by thesentinels** [comment]
:metal: I am sending your branch against master to our acceptance environment.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started the deployment of acceptance](https://travis-ci.org/habitat-sh/habitat/builds/289413252).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports acceptance has been deployed](https://travis-ci.org/habitat-sh/habitat/builds/289413252).

I love new and shiny websites!

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Comment by christophermaier** [comment]
@thesentinels acceptance

**Comment by thesentinels** [comment]
:metal: I am sending your branch against master to our acceptance environment.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started the deployment of acceptance](https://travis-ci.org/habitat-sh/habitat/builds/289423189).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports acceptance has been deployed](https://travis-ci.org/habitat-sh/habitat/builds/289423189).

I love new and shiny websites!

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Comment by christophermaier** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/289428268).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/289428268).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**